ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -642,7 +643,22 @@ GroupInfoReply getGroupInfo(GroupInfoRequest request) {
     final RaftConfigurationProto conf =
         LogProtoUtils.toRaftConfigurationProtoBuilder(getRaftConf()).build();
     return new GroupInfoReply(request, getCommitInfos(), getGroup(), getRoleInfoProto(),
-        dir.isHealthy(), conf);
+        dir.isHealthy(), conf, getLogInfo());
+  }
+
+  LogInfoProto getLogInfo(){
+    long currentTerm = getInfo().getCurrentTerm();
+    LogInfoProto.Builder logInfoBuilder = LogInfoProto.newBuilder();
+    logInfoBuilder.setApplied(TermIndex.valueOf(currentTerm, getInfo().getLastAppliedIndex()).toProto());
+    logInfoBuilder.setCommitted(TermIndex.valueOf(currentTerm, getRaftLog().getLastCommittedIndex()).toProto());
+    logInfoBuilder.setLastEntry(getRaftLog().getLastEntryTermIndex().toProto());
+    if (getStateMachine().getLatestSnapshot() != null) {
+      logInfoBuilder.setLastSnapshot(
+          TermIndex.valueOf(currentTerm,
+              getStateMachine().getLatestSnapshot().getIndex())
+              .toProto());
+    }
```

Comment:
Don't call the same method (`getLatestSnapshot()` in this case) multiple times unless it is returning a constant (i.e. calling `getStateMachine()` multiple times is okay):
1. They may return different values.
2. It may be expensive.

Also, use `snapshot.getTermIndex()`:

After all the changes, it should look like:
```java
  LogInfoProto getLogInfo(){
    final RaftLog log = getRaftLog();
    final LogInfoProto.Builder logInfoBuilder = LogInfoProto.newBuilder()
        .setApplied(getStateMachine().getLastAppliedTermIndex().toProto())
        .setCommitted(log.getTermIndex(log.getLastCommittedIndex()).toProto())
        .setLastEntry(log.getLastEntryTermIndex().toProto());

    final SnapshotInfo snapshot = getStateMachine().getLatestSnapshot();
    if (snapshot != null) {
      logInfoBuilder.setLastSnapshot(snapshot.getTermIndex().toProto());
    }
    return logInfoBuilder.build();
  }
```

Reply:
Thanks Nicholas for the pointer!

Reply:
updated!